services/user-settings-service.ts

```diff
@@ -22,44 +30,67 @@ export class UserSettingsServiceBase implements IUserSettingsService {
 	}
 
 	public async removeSetting(key: string): Promise<void> {
-		await this.loadUserSettingsFile();
+		const action = async () => {
+			await this.loadUserSettingsFile();
+
+			delete this.userSettingsData[key];
+			await this.saveSettings();
+		};
+
+		return this.executeActionWithLock(action);
 
-		delete this.userSettingsData[key];
-		await this.saveSettings();
 	}
 
-	public async saveSettings(data?: any): Promise<void> {
-		await this.loadUserSettingsFile();
-		this.userSettingsData = this.userSettingsData || {};
+	private async executeActionWithLock(action: () => Promise<any>): Promise<any> {
```

Comment:
I believe this could be generic:
`private async executeActionWithLock<T>(action: () => Promise<T>): Promise<T> {`
that way we'd be able to get a little bit more out of the static type checking